prow/tide/history/history.go

```diff
@@ -0,0 +1,159 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package history
+
+import (
+	"encoding/json"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// FE should support:
+// - filter by repo (?branch) (?pr)
+// - filter by author
+// - filter by action
+// - filter by errored state
+// - filter 'before time t'
+// - no-filter. i.e. all actions tide has taken in order.
+
+//
+// ???	Should `History` use a single `*recordLog` instead of per pool?
+//				- This would ensure that the views that look at multiple pools are not 'missing' events in some pools.
+//				- This means many more recent records are saved
+//					- better for active repos so we can increase history time window
+//				- Also means that non-recent records are dropped...
+//					- bad for inactive repos as we may GC most/all the history.
+//
+
+// Mock out time for unit testing.
+var now = time.Now
+
+type History struct {
+	logs map[string]*recordLog
+	sync.Mutex
```

Comment:
is there a reason we chose to use one mutex here instead of making recordLog concurrent-safe?

Reply:
We could use RW mutexes at both levels, but thats overkill (there is at most one record added per pool per sync loop and in practice much less so this lock will never see contention). We need some kind of top level mutex either way so that we can add new pools to the map.